src/Nethermind/Nethermind.Serialization.FluentRlp/RlpWriter.cs

```diff
@@ -0,0 +1,203 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers.Binary;
+using System.Numerics;
+using System.Runtime.InteropServices;
+
+namespace Nethermind.Serialization.FluentRlp;
+
+public delegate void RefRlpWriterAction<in TContext>(ref RlpWriter arg, TContext value)
+    where TContext : allows ref struct;
+
+public delegate void RefRlpWriterAction(ref RlpWriter arg);
+
+public ref struct RlpWriter
+{
+    private const bool LengthMode = false;
+    private const bool ContentMode = true;
+
+    private bool _mode;
+
+    public int Length { get; private set; }
+
+    private byte[] _buffer;
```

Comment:
We often write into Netty arena-based buffers to avoid allocations, would be good to support that.